src/drivers/imx/esai.c

```diff
@@ -426,6 +426,17 @@ static int esai_get_fifo(struct dai *dai, int direction, int stream_id)
 	}
 }
 
+static int esai_get_hw_params(struct dai *dai,
+			      struct sof_ipc_stream_params *params,
+			      int dir)
+{
+	/* ESAI only currently supports these parameters */
```

Comment:
Do you want to warn here if the user requests other values ?

Reply:
I'm not sure where the input parameters should be taken from, and also certain things (sample width) can be adjusted via the current parameter propagation -- host component can adjust sample width as far as I know and thus support any sample width host side.

I think on SSP the input comes from prior DAI configuration (it's customizable). In this case there is no input data to customize how the DAI works. As such, I am just forcing these parameters.

Is this a correct understanding?

Reply:
Clocking, masters and HW format come from topology and pipeline build time whilst rate , channels and sample size would come from PCM hw params at runtime.  It's really up to you based on the capabilities of the HW.

Reply:
Well given the fact that clocking cannot be done from within firmware on any ARM platform I am forced to select the channel count, sample width and sample rate to be a fixed one according to Linux-side configuration of the codec. So, until a protocol is developed for DSP to request from ARM to configure clocking and masters, I will have to force fixed hardware parameters.

Hardware itself would be capable of other sample rates and widths (still only signed integer samples, no support for float or other formats) but that requires manually configuring PLLs which are controlled exclusively by the ARM side of things. For other formats I think pipeline must contain converters such as (A)SRC so they are brought in line with what we have.

Until we can figure out a way to properly control clocking so that other sample rates, channel widths etc are possible we will limit ourselves to this configuration. The topology is written currently to agree with the available capabilities.